### PR TITLE
The pull request that changed Nukkit generator forever!

### PR DESCRIPTION
**Pull request by Niall7459**
This is a very very big step foward in our generator!

Thanks to @Creeperface01  and @NycuRO I have been able to port their nukkit plugin (BiomeMaster) to the nukkit core.
https://github.com/NycuRO/BiomeMaster
This PR should be ready to merge but there may be a few unfinished things.

**Adds:**
-Jungle biome
-Roofed forest biomes
-Savanna biome
-Mushroom islands

-New trees for biomes

**TODO:**
- Check the other biomes are appearing (very hard to find some of them)
- Add the trees to generate when planting sapling.
**- Possible change to fit different structure?**
Screenshot:
screen shot 2016-11-03 at 16 17 52
screen shot 2016-11-03 at 16 42 59

### CODE REVIEW

File: src/main/java/cn/nukkit/level/generator/Normal.java [hunk 1]
```diff
@@ -110,25 +160,28 @@ public void init(ChunkManager level, NukkitRandom random) {
         this.noiseBaseGround = new Simplex(this.nukkitRandom, 4F, 1F / 4F, 1F / 64F);
         this.noiseRiver = new Simplex(this.nukkitRandom, 2F, 1F, 1F / 512F);
         this.nukkitRandom.setSeed(this.level.getSeed());
-        this.selector = new BiomeSelector(this.nukkitRandom, Biome.getBiome(Biome.OCEAN));
+        this.selector = new BiomeSelector(this.nukkitRandom, Biome.getBiome(Biome.FOREST));
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/Normal.java]: @Niall7459 This is wrong . Is Plains. I need to change to BiomeMaster.

File: src/main/java/cn/nukkit/level/generator/biome/BiomeSelector.java [hunk 2]
```diff
@@ -26,44 +29,65 @@ public BiomeSelector(NukkitRandom random, Biome fallback) {
...
         }
+    return Biome.OCEAN;
     }
-
+    
     public void recalculate() {
         this.map = new int[64 * 64];
-        for (int i = 0; i < 64; ++i) {
-            for (int j = 0; j < 64; ++j) {
+        for(int i = 0; i < 64; ++i) {
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/biome/BiomeSelector.java]: @Niall7459 Spaces here ,please

File: src/main/java/cn/nukkit/level/generator/object/dev/mushroom/DevBigMushroom.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
+package cn.nukkit.level.generator.object.dev.mushroom;
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/object/dev/mushroom/DevBigMushroom.java]: @Niall7459 Why dev...?

File: src/main/java/cn/nukkit/level/generator/object/dev/tree/DevDarkOakTree.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
+package cn.nukkit.level.generator.object.dev.tree;
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/object/dev/tree/DevDarkOakTree.java]: Why dev again? :/

File: src/main/java/cn/nukkit/level/generator/object/dev/tree/DevJungleBigTree.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
+package cn.nukkit.level.generator.object.dev.tree;
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/object/dev/tree/DevJungleBigTree.java]: I hate "dev" folder...

File: src/main/java/cn/nukkit/level/generator/object/dev/tree/DevJungleTree.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
+package cn.nukkit.level.generator.object.dev.tree;
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/object/dev/tree/DevJungleTree.java]: Again Dev ...

File: src/main/java/cn/nukkit/level/generator/object/dev/tree/DevSavannaTree.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
+package cn.nukkit.level.generator.object.dev.tree;
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/object/dev/tree/DevSavannaTree.java]: Again dev..

File: src/main/java/cn/nukkit/level/generator/object/dev/tree/DevSwampTree.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
+package cn.nukkit.level.generator.object.dev.tree;
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/object/dev/tree/DevSwampTree.java]: Dev.....

File: src/main/java/cn/nukkit/level/generator/object/dev/mushroom/DevBigMushroom.java (hunk 3 above)
Comment (Niall7459) [inline src/main/java/cn/nukkit/level/generator/object/dev/mushroom/DevBigMushroom.java]: Because its a in-dev thing

Comment (Niall7459) [inline src/main/java/cn/nukkit/level/generator/object/dev/mushroom/DevBigMushroom.java]: I would want the nukkit team to approve it before editing the structure

**Comment by NycuRO** [comment]
@Niall7459 I'm preffer to create PR when plugin is complete..

**Comment by Creeperface01** [comment]
hasn't jungle wird grass color?

**Comment by Niall7459** [comment]
@Creeperface01 No thats swampcolor

**Comment by MagicDroidX** [comment]
OH MY GOD
What I have to say is that GOOOOOOOOOD JOB!
I am looking to this pr

**Comment by MagicDroidX** [comment]
Is there possibility to implement smooth biome color change?

**Comment by Niall7459** [comment]
@MagicDroidX Anything for the nukkit project!

**Comment by Niall7459** [comment]
@MagicDroidX It doesn't really fit the original structure very well
TODO:

- [ ] Smooth biome colour
- [ ] Make it fit nukkit structure
- [ ] Make it work with saplings!

**Comment by Niall7459** [comment]
I will also confirm other biomes are working and add screens

**Comment by SuperMaXAleX** [comment]
Awesome. If you can, add Ocean Temples !

**Comment by NycuRO** [comment]
@MagicDroidX You can use BiomeMaster. Is created specially for Nukkit,not for other softwares. #Nukkit_Great

But is bug with generation . Need change method with rainfall and temperature because idk how values need add for generate all biomes.

**Comment by NycuRO** [comment]
@Creeperface01 Watch generation.txt in BiomeMaster . If i'm place this railfall and temperatures not work.

**Comment by Creeperface01** [comment]
@NycuRO https://github.com/erich666/Mineways/blob/master/Win/biomes.cpp

**Comment by NycuRO** [comment]
@Niall7459 Please update PR from plugin.

**Comment by Pub4Game** [comment]
Is it worth to add?

**Comment by yungtechboy1** [comment]
Yea I think so... But setting elevation is not implemented

**Comment by Creeperface01** [comment]
It must be... Look at plains biome and mountains...

**Comment by yungtechboy1** [comment]
@Creeperface01 
https://github.com/Nukkit/Nukkit/blob/1.0.0/src/main/java/cn/nukkit/level/generator/Normal.java#L197

That genyHeight value is never changed by chunk, So its just the seed.

**Pull request closed without merging**